File: ONCE_A_WEEK/WEEK6/곽태웅/23.py
from collections import Counter

def solution(genres, plays):
    play_list = [{'genre': genres[i], 'play': plays[i], 'index': i} for i in range(len(genres))]
    
    sorted_list = sorted(play_list, key=lambda x: (-x['play'], x['index']))

    genres_counter = Counter()
    for genre, play in zip(genres, plays):
        genres_counter[genre] += play

    sorted_genres = genres_counter.most_common()

    answer = []
    
    for genre, _ in sorted_genres:
        count = 0
        for song in sorted_list:
            if song['genre'] == genre:
                count += 1
                if count <= 2:
                    answer.append(song['index'])
                else:
                    break

    return answer

solution(["classic", "pop", "classic", "classic", "pop"], 	[500, 600, 150, 800, 2500])

# 너무 유지보수하기 어렵고, 복잡한 코드가 아닐까...
# 결론적으로 속도는 O(n log n) 수준이니, 이보다 간단하게 구현할 방법이 필요할 것 같다

File: ONCE_A_WEEK/WEEK6/곽태웅/23.js
// 절대 자바스크립트로는 코테 보지 말아야 하는 이유

function solution(genres, plays) {
    let l = genres.length;
        map = new Map();
        arr = Array.from([...new Set(genres)]);
    for (let i = 0; i < l; i++) {
        let g = arr[arr.indexOf(genres[i])];
        map.set(g, (map.get(g) || 0) + plays[i]);
    }
    arr = [...map.entries()].sort((a, b) => b[1] - a[1]);
    let genres_map = new Map;
    for (let i = 0; i < l; i++) {
        let g = genres[i];
        genres_map.set(g, (genres_map.get(g) || 0) + 1);
    }
    let genres_obj = [];
    for (let i = 0; i < l; i++) {
        genres_obj[i] = {genres: genres[i], index: i, plays: plays[i], genresCount: genres_map.get(genres[i])};
    }
    genres_obj.sort((a, b) => b.plays - a.plays);
    for (let i = 0; i < arr.length; i++) {
        genres_obj.sort((a, b) => {
            if (a.genres === arr[i][0] && b.genres !== arr[i][0]) {
                return -1;
            } else if (a.genres !== arr[i][0] && b.genres === arr[i][0]) {
                return 1;
            } else {
                return 0;
            }
        })
    }
    genres_obj.reverse();
    genres_obj.push({genres: 'genre', index: 0, plays: 0});
    genres_obj.push({genres: 'genre', index: 0, plays: 0});

    let result = [];
    arr = [];
    let j = 0;
    for (let i = 0; i < l; i++) {
        if (genres_obj[i].genresCount == 1) {
            result.push(genres_obj[i].index);
        } else if (genres_obj[i].genres !== genres_obj[i + 2].genres) {
            j++;
            result.push(genres_obj[i].index);
        } else {
            result.push(-1);
        }
        if (j >= 2) {
            j = 0;
            let r = result[i - 1];
            result[i - 1] = result[i];
            result[i] = r;
        }
    }
    result = result.filter((v) => v >= 0);
    return result;
}